library/tiqr/Tiqr/Message/APNS2.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+use Pushok\AuthProvider;
+use Pushok\Client;
+use Pushok\Notification;
+use Pushok\Payload;
+use Pushok\Payload\Alert;
+
+/** @internal includes */
+require_once('Tiqr/Message/Abstract.php');
+
+/**
+ * Apple Push Notification Service message class for the HTTP/2 api.push.apple.com APNs API.
+ */
+class Tiqr_Message_APNS2 extends Tiqr_Message_Abstract
+{
+    /**
+     * Send message.
+     */
+    public function send()
+    {
+        $version_info = curl_version();
+        if ($version_info['features'] & CURL_VERSION_HTTP2 == 0) {
+            throw new RuntimeException('APNS2 requires HTTP/2 support in curl');
+        }
+
+        // Get the UID from the client certificate we use for authentication, this
+        // is set to the bundle ID.
+        $options=$this->getOptions();
+        $cert_filename = $options['apns.certificate'];
+        $cert_file_contents = file_get_contents($cert_filename);
+        if (false === $cert_file_contents) {
+            throw new RuntimeException(
+                sprintf('Error reading APNS client certificate file: "%s"', $cert_filename)
+            );
+        }
+
+        $cert=openssl_x509_parse( $cert_file_contents );
+        if (false === $cert) {
+            throw new RuntimeException('Error parsing APNS client certificate');
+        }
+        $bundle_id = $cert['subject']['UID'] ?? NULL;
+        if (NULL === $bundle_id) {
+            throw new RuntimeException('No uid found in the certificate subject');
+        }
+        $this->logger->info(sprintf('Setting bundle_id to "%s" based on UID from certificate', $bundle_id));
+        $this->logger->info(
+            sprintf('Authenticating using certificate with subject "%s" valid until "%s"',
+                $cert['name'],
+                date(DATE_RFC2822, $cert['validTo_time_t'])
+            )
+        );
+
+        $authProviderOptions = [
+            'app_bundle_id' => $bundle_id, // The bundle ID for app obtained from Apple developer account
+            'certificate_path' => $cert_filename,
+            'certificate_secret' => null // Private key secret
+        ];
+
+        $authProvider = AuthProvider\Certificate::create($authProviderOptions);
+
+        // Create the push message
+        $alert=Alert::create();
+        $alert->setBody($this->getText());
+        // TODO: Set title?
```

Comment:
The `Tiqr_Message_Abstract` does not implement a getter or a field for the title of the push notification message. Is this something you'd like to implement in this PR give you left a todo comment? Or is this a marker for future development. In that case my personal preference is to remove the TODO signs and move them into Pivotal.

/Nitpick

Reply:
https://www.pivotaltracker.com/story/show/183081723